### PR TITLE
Add security considerations (close #149)

### LINKED ISSUE #149: Security considerations
<https://mailarchive.ietf.org/arch/msg/asdf/pLqaOGTs0ixMMM33XWCcJcPDLWo>

### PR DESCRIPTION
**Pull request by cabo**
Close #149

### CODE REVIEW

File: sdf.md [hunk 1]
```diff
@@ -1816,10 +1818,72 @@ The initial set of registrations is described in {{sdftype-r}}.
...
+Some wider security considerations applicable to Things are discussed
+in {{-seccons}}.
+{{Section 5 of -cddl}} gives an overview over security considerations
+that arise when formal description techniques are used to govern
+interoperability; analogs of these security considerations can apply
+to SDF.
+
+The security considerations of underlying building blocks such as
+those detailed in {{Section 10 of -utf8}} apply.
+SDF uses JSON as a representation language; {{-json}} for a number of
```
Comment (akeranen) [inline sdf.md:1830]: Some punctation missing here? Or 
> For a number of cases {{-json}} indicates ...

File: sdf.md [hunk 2]
```diff
@@ -1816,10 +1818,72 @@ The initial set of registrations is described in {{sdftype-r}}.
...
+performance-related availability issues when the attacker can control
+the input, see {{Section 4.1 of -jsonpath}} for a brief discussion.
+
+SDF may be used in two processes that are often security relevant:
+*validation* of data that is intended to be described by SDF models, and
+*augmentation* of these data with information obtained from the SDF
+models that apply.
+
+Implementations need to ascertain the provenance and applicability of
+the SDF models they employ operationally in these ways.
```
Comment (akeranen) [inline sdf.md:1848]: Where "these ways" refers to would be more clear without the paragraph break above

File: sdf.md [hunk 3]
```diff
@@ -1816,10 +1818,72 @@ The initial set of registrations is described in {{sdftype-r}}.
...
+Essentially, this process needs to undergo the same care and scrutiny
+as any other introduction of source code into a build environment; the
+possibility of supply-chain attacks on the modules imported needs to
+be considered.
+
+Specifically, implementations that rely on model-based input
+validation for enforcing certain properties of the data structure
+ingested (which, if not validated, could lead to malfunctions such as
+crashes and remote code execution) need to be particularly careful
+about the data models they apply, including their provenance and
```
Comment (akeranen) [inline sdf.md:1861]:
```suggestion
about the data models they apply. This includes their provenance and
```

Suggesting breaking long sentence into two for readability

File: sdf.md [hunk 4]
```diff
@@ -1816,10 +1818,72 @@ The initial set of registrations is described in {{sdftype-r}}.
...
+potential changes of these properties that upgrades to the referenced
+modules may (inadvertently or as part of an attack) cause.
+More generally speaking, implementations should strive to be robust
+against expected and unexpected limitations of the model-based input
+validation mechanisms and their implementations.
+
+Similarly, implementations that rely on model-based augmentation may
```
Comment (akeranen) [inline sdf.md:1868]: Perhaps useful to add some words to explain what we mean why "model-based augmentation"?

**Review by akeranen — APPROVED** [review]
LGTM! Some editorial suggestions inline.

File: sdf.md (hunk 3 above)
Comment (cabo) [reply sdf.md:1861]: I broke it up in a different place.

File: sdf.md (hunk 2 above)
Comment (cabo) [reply sdf.md:1848]: I think the paragraph break is useful, so I changed "these ways" into " such security relevant ways".

File: sdf.md (hunk 1 above)
Comment (cabo) [reply sdf.md:1830]: The sequence I used is valid but may indeed be surprising, so I changed this along the lines you suggested.

File: sdf.md (hunk 4 above)
Comment (cabo) [reply sdf.md:1868]: The fourth paragraph was intended to do that.  I added "model-based" twice.

File: sdf.md (hunk 1 above)
Comment (mjkoster) [reply sdf.md]: ... representation language. For a number of cases, {{-json}} indicates that ...

**Review by mjkoster — APPROVED** [review]

**Review by akeranen — APPROVED** [review]
Fixes LGTM. Thanks!

File: sdf.md [hunk 5]
```diff
@@ -1816,10 +1818,73 @@ The initial set of registrations is described in {{sdftype-r}}.
...
+
+The security considerations of underlying building blocks such as
+those detailed in {{Section 10 of -utf8}} apply.
+SDF uses JSON as a representation language; for a number of
+cases {{-json}} indicates that implementation behavior for certain constructs
+allowed by the JSON grammar is unpredictable.
+Implementations need to be robust against invalid or unpredictable
+cases on input, preferably by rejecting input that is invalid or
+that would lead to unpredictable behavior, and need to avoid generating
+these cases on output.
```
Comment (mcr) [inline sdf.md:1836]: What should implementations do when they see invalid or unpredictable input?  Should the entire message be dropped (do we have a clear error indication in all mappings?  Do we have mappings for errors?), or just the construct which was invalid, assuming that the format allows for that?
It also might be useful to distinguish code that runs at development time from code that might be generated.

**Review by mcr — APPROVED** [review]

File: sdf.md (hunk 5 above)
Comment (cabo) [reply sdf.md:1836]: I think this is very hard to answer in general.
There are several instances in the text that discuss validity, but we don't have a complete validity model.
What do you do when you get an ABNF spec that is invalid?
I'll merge this PR without having fully addressed your comment; if you think we need to develop a validity model first before going ahead with this document, please add another issue.

**Pull request merged**